### PR TITLE
[WT-75] Set up ticket exchanges

### LINKED ISSUE #75: Setup ticket exchanges
Currently, a ticket exchange could work by deleting the user from the doorlist and adding them to a new showing instance. It would be awesome if there was a system where you can click a swap button and exchange the tickets without the extra effort.

### PR DESCRIPTION
**Pull request by benfaustin95**
### Description
Add the ability to do order item exchanges as well as partial refunds. 

Database updates:
- Combined `orders` and `refunds` into one table so that `refunditems` can be netted with `orderitems` on a single order. 
- Added `payment_intent` and `order_payment_intent` tables so that a single `payment_intent` can be spread across multiple orders as items are exchanged/refunded.
- Added a `temporarystore` table so that we do not have to rely on stripe metadata (which is limited in size) to handle tasks that need to be completed once an order has finished processing.

Backend Updates: 
- Centralized checkout logic for all order and refund methods.
- Updated server and seeding logic to handle the new database structure/refund process.
- Added price validation for non-admin orders.
- Added logic to allocate discount among tickets, this is probably the wrong logic but is more of placeholder as exchanges on orders with discounts didn't really work without accounting for the ticket level discount. 

Frontend Updates:

- Added UI to allow for order exchanges. 

image

### Risks
It is very possible that the implementation logic around the new order structure is incorrect, or that I missed an edge case when testing. 

### Validation
Manual testing

### Issue
closes #75 

### Operating System
M1 Mac

### CODE REVIEW

**Review by henrykaus — COMMENTED** [review]
This is a fantastic piece of functionality, and it works really well. I think the way you chose to handle refunds or charging extra makes a lot of sense. Due to the size of the PR, I mostly just did manual testing of the feature and here is a little bit of feedback:

In the Cart Section of the page, it might make more sense (if possible) to have 2 sections, one like `Items to Return` and `Items to Exchange` where the the customer's already-purchased items appear in the former and the items-to-exchange-for appear in the later. In this way, we could avoid using the financial way of displaying cost values (parens vs no parens), and it might make it more immediately user friendly to have items grouped in this way.

Some Bugs I Noticed:
* If you set the price of tickets you're exchanging for to `$0` but have fees on those items, the user is not charged the fee. Maybe have the UI set the fee to 0 and disable it if the price is 0
* When applying a discount code to an exchange (like exchanging 2 $20 tickets for 2 other $20 tickets), there can be a negative total _as a result of_ the discount code, but discount codes should not make the order go below 0.

File: client/src/components/Ticketing/mainpage/SubscriptionPurchasing/QuantityInputControl.tsx [hunk 1]
```diff
@@ -1,38 +1,44 @@
...
+    quantity: string;
+    icon?: string;
+  };
 }
 
 export const QuantityInputControl = (props: QuantityInputControl) => {
-  const {field, quantityAvailable} = props;
-  const {setFieldValue} = useFormikContext();
-
+  const {field, quantityAvailable, decrement, increment, styles, disabled = false} = props;
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/mainpage/SubscriptionPurchasing/QuantityInputControl.tsx:21]: Fixed

File: client/src/components/Ticketing/ticketingmanager/TicketExchanges/CustomerDisplay.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  return (
+    <div className='p-4 grid grid-cols-4 w-fit h-fit place-self-center'>
+      <p className='text-lg font-semibold col-span-3'>
+        {customer.firstname} {customer.lastname}
+      </p>
+      <div className='col-span-1 flex justify-end'>
+        <FormButton
+          disabled={stage === 'checkout'}
+          testID='remove-customer-from-order'
+          title='remove customer-from-order'
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/ticketingmanager/TicketExchanges/CustomerDisplay.tsx]: Fixed

File: client/src/components/Ticketing/ticketingmanager/TicketExchanges/TicketExchangeItems.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  return (
+    <div className='w-full grid grid-cols-12'>
+      <section className='col-span-12 min-[1280px]:col-span-8 p-3'>
+        <header className='flex flex-row justify-between mb-3'>
+          <Tabs
+            value={current}
+            onChange={(_, newValue) => setCurrent(newValue)}
+            variant='scrollable'
+          >
+            <Tab label='ORDERS' value={1} disabled={!customer} />
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/ticketingmanager/TicketExchanges/TicketExchangeItems.tsx]: Fixed

File: server/src/controllers/eventController.service.ts [hunk 4]
```diff
@@ -129,12 +79,32 @@ export const checkout = async (
...
+  metadata,
+}: CreateStripeCheckoutSessionArguments) => {
   const expire = Math.round((new Date().getTime() + 1799990) / 1000);
+  const discounts = [];
+
+  if (discount.code != '' && discount.amountOff) {
+    discounts.push({coupon: (await createStripeCoupon(discount)).id});
+  }
+  if (refundCredit && refundCredit > 0) {
+    discounts.push({coupon: (await createStripeCoupon({code: 'refund_credit', amountOff: refundCredit})).id});
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply server/src/controllers/eventController.service.ts]: Fixed

**Comment by benfaustin95** [comment]
> This is a fantastic piece of functionality, and it works really well. I think the way you chose to handle refunds or charging extra makes a lot of sense. Due to the size of the PR, I mostly just did manual testing of the feature and here is a little bit of feedback:
> 
> In the Cart Section of the page, it might make more sense (if possible) to have 2 sections, one like `Items to Return` and `Items to Exchange` where the the customer's already-purchased items appear in the former and the items-to-exchange-for appear in the later. In this way, we could avoid using the financial way of displaying cost values (parens vs no parens), and it might make it more immediately user friendly to have items grouped in this way.
> 
> Some Bugs I Noticed:
> 
> * If you set the price of tickets you're exchanging for to `$0` but have fees on those items, the user is not charged the fee. Maybe have the UI set the fee to 0 and disable it if the price is 0
> * When applying a discount code to an exchange (like exchanging 2 $20 tickets for 2 other $20 tickets), there can be a negative total _as a result of_ the discount code, but discount codes should not make the order go below 0.
updated

**Review by henrykaus — COMMENTED** [review]
Everything looks good. I can approve after this last issue is fixed:

In ticket exchanges, when you select a showing for an event, and then select a ticket type -- but then reset the showing to the `Select Showing` option the page crashes.

**Comment by benfaustin95** [comment]
> Everything looks good. I can approve after this last issue is fixed:
> 
> In ticket exchanges, when you select a showing for an event, and then select a ticket type -- but then reset the showing to the `Select Showing` option the page crashes.

Fixed

**Review by henrykaus — APPROVED** [review]
Looks good! Thank you for implementing this beast of a feature!

**Pull request merged**